adapters/eplanning/eplanning.go

```diff
@@ -8,86 +8,109 @@ import (
-}
+	for i := 0; i < totalImps; i++ {
+		source, err := verifyImp(&request.Imp[i])
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
 
-type EPlanningAdUnit struct {
-	Id         string
-	Bidfloor   float64
-	Instl      int8
-	SspSpaceId int `json:"ssp_espacio_id,omitempty"`
-	Video      *openrtb.Video
-	Banner     *openrtb.Banner
-}
+		if _, ok := sourceMapper[source]; !ok {
+			sourceMapper[source] = make([]int, 0, totalImps-i)
+		}
 
-func (adapter *EPlanningAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
-	ePlanningRequest, errors := openRtbToEPlanningRequest(request)
-	if len(ePlanningRequest.adUnits) == 0 {
-		return nil, errors
+		sourceMapper[source] = append(sourceMapper[source], i)
 	}
 
-	reqJSON, err := json.Marshal(request)
-	if err != nil {
-		errors = append(errors, err)
+	totalRequests := len(sourceMapper)
+
+	if totalRequests == 0 {
 		return nil, errors
 	}
 
-	requestData := adapters.RequestData{
-		Method: "POST",
-		Uri:    adapter.URI,
-		Body:   reqJSON,
+	requests := make([]*adapters.RequestData, 0, totalRequests)
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	headers.Add("Accept", "application/json")
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(request.Device.DNT)))
 	}
 
-	requests := []*adapters.RequestData{&requestData}
+	imps := request.Imp
 
-	return requests, errors
-}
+	for source, impIds := range sourceMapper {
+		request.Imp = request.Imp[:0]
 
-func openRtbToEPlanningRequest(request *openrtb.BidRequest) (*EPlanningRequest, []error) {
+		for i := 0; i < len(impIds); i++ {
+			request.Imp = append(request.Imp, imps[impIds[i]])
+		}
 
-	adUnits := make([]*EPlanningAdUnit, 0, len(request.Imp))
-	errors := make([]error, 0, len(request.Imp))
-	for _, imp := range request.Imp {
-		var params openrtb_ext.ExtImpEPlanning
-		err := json.Unmarshal(imp.Ext, &params)
+		reqJSON, err := json.Marshal(request)
 		if err != nil {
 			errors = append(errors, err)
-			continue
+			return nil, errors
 		}
-		ePlanningAdUnit := EPlanningAdUnit{
-			Id:         imp.ID,
-			Bidfloor:   imp.BidFloor,
-			Instl:      imp.Instl,
-			Video:      imp.Video,
-			Banner:     imp.Banner,
-			SspSpaceId: params.SspSpaceId,
+
+		requestData := adapters.RequestData{
+			Method:  "POST",
+			Uri:     adapter.URI + fmt.Sprintf("/%s", source),
+			Body:    reqJSON,
+			Headers: headers,
 		}
-		adUnits = append(adUnits, &ePlanningAdUnit)
+
+		requests = append(requests, &requestData)
+	}
+
+	return requests, errors
+}
+
+func verifyImp(imp *openrtb.Imp) (string, error) {
+	// We currently only support banner impressions
+	if imp.Native != nil || imp.Audio != nil || imp.Video != nil {
```

Comment:
Multi-format Imps may be a thing some day. You might want to bid on the banner option of a multi-format. Also if a new type is developed some day, you will have to add it to your black list. You might want to consider simply rejecting all that have a non nil Banner object.

Reply:
multi-format Imps _are_ a thing _today_ :)